subsys/bluetooth/controller/hci/hci.c

```diff
@@ -3624,33 +3630,83 @@ int hci_acl_handle(struct net_buf *buf, struct net_buf **evt)
 #endif /* CONFIG_BT_CONN */
 
 #if CONFIG_BT_CTLR_DUP_FILTER_LEN > 0
-static inline bool dup_found(struct pdu_adv *adv)
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+static void store_adi(int i, struct pdu_adv_adi *adi)
+{
+	if (adi) {
+		memcpy(&dup_filter[i].adi, adi, sizeof(*adi));
+	} else {
+		memset(&dup_filter[i].adi, 0, sizeof(*adi));
+	}
+}
+#endif /* CONFIG_BT_CTLR_ADV_EXT */
+
+static bool dup_found(uint8_t adv_type, uint8_t addr_type, uint8_t *addr,
+		      uint8_t adv_mode, struct pdu_adv_adi *adi,
+		      uint8_t data_status)
 {
 	/* check for duplicate filtering */
 	if (dup_count >= 0) {
 		int i;
 
 		for (i = 0; i < dup_count; i++) {
-			if (!memcmp(&adv->adv_ind.addr[0],
-				    &dup_filter[i].addr.a.val[0],
+			if (!memcmp(addr, &dup_filter[i].addr.a.val[0],
 				    sizeof(bt_addr_t)) &&
-			    adv->tx_addr == dup_filter[i].addr.type) {
+			    (addr_type == dup_filter[i].addr.type)) {
+
+				if (!(dup_filter[i].mask & BIT(adv_type))) {
+					/* report different adv types */
+					dup_filter[i].mask |= BIT(adv_type);
+
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+					dup_filter[i].adv_mode = adv_mode;
+					dup_filter[i].data_cmplt = !data_status;
+					store_adi(i, adi);
+
+					return false;
+				} else if (dup_filter[i].adv_mode != adv_mode) {
+					/* report different adv mode */
+					dup_filter[i].adv_mode = adv_mode;
```

Comment:
Should adv_mode use a union of reported values like it is done for adv_type?

Reply:
No. As SID is set by the application when creating adv set, peer may use the same SID and DID. To be on safer side, will consider changes to adv mode as change to data.